### PR TITLE
Support vllm_decode for multimodal (Qwen3-VL)

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description

**One-sentence Summary**: We added extensible vLLM multimodal preprocessing that runs MaxText’s vision encoder and passes merged text-image embeddings into the existing MaxText decoder:

- Adds Qwen3-VL image decoding through vLLM while keeping the vision encoder and language model in MaxText.
- vLLM/tpu-inference preprocesses images and merges visual and text embeddings before invoking the model, so MaxText receives `decoder_input_embeddings` instead of its native, pre-merge `multimodal_input`.
- Introduces `MultimodalHandler` for model-specific processor registration, tensor conversion, placeholder tokens, and MaxText modality encoding.
- **For a future MaxText-only model**, add a handler and a vLLM-compatible processor in MaxText; no separate vLLM model implementation is required.

**Important Note**: vLLM passes token IDs for text-only requests, but passes complete, premerged text and multimodal embeddings for multimodal requests ([reference](https://screenshot-v2.corp.google.com/27ci3vk2mem0g)). The latter enter through `decoder_input_embeddings` to avoid embedding and merging them again.

# Tests

```
python -m maxtext.inference.vllm_decode src/maxtext/configs/base.yml run_name=vllm_decode_qwen3_vl model_name=qwen3-vl-2b load_parameters_path=/home/hengtaoguo_google_com/projects/checkpoints/qwen3-vl-2b/unscanned/2026-07-01-11-55/0/items tokenizer_path=Qwen/Qwen3-VL-2B-Instruct prompt=\"Describe\ this\ image:\ \" image_path=/home/hengtaoguo_google_com/projects/maxtext/tests/assets/test_image.jpg max_prefill_predict_length=1024 max_target_length=2048 per_device_batch_size=1 scan_layers=false weight_dtype=bfloat16 max_num_batched_tokens=16384 use_chat_template=true vllm_hf_overrides=\{\"architectures\":\[\"MaxTextForCausalLM\"\]\}
```

```
Rendering prompts: 100%|████████████████████████| 1/1 [00:09<00:00,  9.43s/it]
Processed prompts: 100%|█| 1/1 [00:21<00:00, 21.93s/it, est. speed input: 16.1
I0727 04:42:04.155640 140668460825728 vllm_decode.py:229] Prompt: <|im_start|>user
<|vision_start|><|image_pad|><|vision_end|>Describe this image: <|im_end|>
<|im_start|>assistant
, Generated text: This is a panoramic photograph of the Seattle city skyline on a clear, sunny day.

The most prominent feature is the **Space Needle**, a distinctive observation deck that stands out in the center of the image. It is surrounded by a dense collection of modern skyscrapers and high-rise buildings that form the city's downtown core and business district. The buildings have a variety of architectural styles and glass facades, giving the skyline a dynamic and vibrant appearance.

In the foreground, the city is populated with a mix of residential buildings, including apartment complexes and smaller housing units, nestled among lush green trees and parks. These natural elements provide a contrast to the concrete and steel of the city's architecture.

In the background, a range of mountains is visible, with snow-capped peaks, suggesting the scenic beauty of the Pacific Northwest. The sky is a brilliant, clear blue, with a few soft, white clouds drifting across it.
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/inference/vllm_decode.py [hunk 1]
```diff
@@ -65,12 +74,18 @@
...
-def build_chat_messages(config: Config) -> list[dict[str, str]]:
+def build_chat_messages(config: Config) -> list[dict[str, Any]]:
   """Builds the chat message list, prepending a system prompt when set."""
   messages = []
   if config.system_prompt:
     messages.append({"role": "system", "content": config.system_prompt})
-  messages.append({"role": "user", "content": config.prompt})
+
+  if config.use_multimodal:
+    content = [{"type": "image"} for _ in config.image_path.split(",")]
```
Comment (dipannita08) [inline src/maxtext/inference/vllm_decode.py:84]: Should null check image_path here?

File: src/maxtext/inference/vllm_decode.py [hunk 2]
```diff
@@ -155,6 +173,11 @@ def decode_with_vllm(config: Config) -> None:
     prompts = [input_with_chat_template]
 
   max_prompt_length = max(len(tokenizer.encode(p)) for p in prompts)
+
+  if config.use_multimodal:
+    images = [Image.open(path.strip()) for path in config.image_path.split(",")]
```
Comment (dipannita08) [inline src/maxtext/inference/vllm_decode.py]: Can we use [load_image_from_path](https://github.com/AI-Hypercomputer/maxtext/blob/2586d60b7467a84eb9ef8caad334f733d34a545e/src/maxtext/multimodal/utils.py#L75) here?

**Review by dipannita08 — APPROVED** [review]
LGTM! Thanks Hengtao!

File: src/maxtext/inference/vllm_decode.py (hunk 1 above)
Comment (hengtaoguo) [reply src/maxtext/inference/vllm_decode.py:84]: Good call, conditioning on `config.image_path` here too.

File: src/maxtext/inference/vllm_decode.py (hunk 2 above)
Comment (hengtaoguo) [reply src/maxtext/inference/vllm_decode.py]: Great suggestion, I changed to reuse `load_image_from_path` here.

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/multimodal.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    # need vLLM's multimodal model modules during adapter import.
+    from vllm.model_executor.models.qwen3_vl import (  # pylint: disable=import-outside-toplevel
+        Qwen3VLDummyInputsBuilder,
+        Qwen3VLMultiModalProcessor,
+        Qwen3VLProcessingInfo,
+    )
+    from vllm.multimodal import MULTIMODAL_REGISTRY  # pylint: disable=import-outside-toplevel
+
+    MULTIMODAL_REGISTRY.register_processor(
+        Qwen3VLMultiModalProcessor,
```
Comment (lydhr) [inline src/maxtext/integration/vllm/maxtext_vllm_adapter/multimodal.py:63]: Will there be any issues when training used MaxText's native processor, then inference uses vLLM native processor?

**Review by lydhr — APPROVED** [review]

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/multimodal.py (hunk 3 above)
Comment (hengtaoguo) [reply src/maxtext/integration/vllm/maxtext_vllm_adapter/multimodal.py:63]: Great question! It should generally be compatible, because vLLM [wraps](https://screenshot-v2.corp.google.com/2g3i2hqcdg0k0) the original Hugging Face Qwen3-VL processor rather than reimplementing it. 

In contrast, maxtext didn't import torch and instead reimplement everything (resize, normalize etc) using pure JAX/numpy. We also had [unit test](https://github.com/AI-Hypercomputer/maxtext/blob/cc79eb8b0172849875f9dd767946113dc9c40d28/tests/unit/qwen3_omni_layers_test.py#L1026) to make sure our processor is equivalent to the original HuggingFace processor.
